app/src/main/java/com/github/se/icebreakrr/model/tags/TagsRepository.kt

```diff
@@ -0,0 +1,250 @@
+
+            tagsCategory.let { category: TagsCategory ->
+              if (!category.subtags.contains(name)) {
+                val updatedSubtags = category.subtags.toMutableList()
+                updatedSubtags.add(name)
+
+                docRef
+                    .update("subtags", updatedSubtags)
+                    .addOnFailureListener { exception: Exception ->
+                      Log.e("TagsRepository", "[addTag] Could not update the document: $exception")
+                      onFailure(exception)
+                    }
+                    .addOnSuccessListener {}
+              } else {
+                Log.d(
+                    "TagsRepository",
+                    "[addTag] Tag '$name' already exists in category '${category.name}'")
+              }
+            }
+          } else {
+            Log.d(
+                "TagsRepository", "[addTag] Document does not exist for category: ${category.name}")
+            onFailure(Exception("Document does not exist"))
+          }
+        }
+  }
+
+  /**
+   * Function that adds a new category to the firestore (as a document) based on a name and a list
+   * of subtags. If the category on firebase already exists, add elements of the subcateries that
+   * are not in the category already present in the firebase to the category on firebase. Else it
+   * just create a new category with the subtags in it. IMPORTANT : the name of the new category
+   * must be put manually after the execution of the function into the enum CategoryString in
+   * model/tags/Tags.kt
+   *
+   * @param onFailure : callback function called if there is an error
+   * @param categoryName : name of the new category
+   * @param subcategories : list of subcategories to have under the name category
+   */
+  fun addCategory(
+      onFailure: (Exception) -> Unit,
+      categoryName: String,
+      subcategories: List<String>,
+      color: String
+  ) {
+    // check if the category is already in the database
+    if (CategoryString.values().any { it.name.equals(categoryName, ignoreCase = true) }) {
+      db.collection(collectionPath)
+          .document(categoryName)
+          .get()
+          .addOnSuccessListener { documentSnapshot: DocumentSnapshot ->
+            var tagsCategory = TagsCategory()
+            firestoreToTags(
+                documentSnapshot,
+                { tagCategoryCallback -> tagsCategory = tagCategoryCallback },
+                { e: Exception ->
+                  Log.e(
+                      "TagsRepository",
+                      "[addCategory] error while converting firebase to Tags : $e")
+                  onFailure(e)
+                })
+            val subtags = tagsCategory.subtags.toMutableList()
+            subtags.addAll(subcategories.filter { it !in subtags })
+            db.collection(collectionPath)
+                .document(categoryName)
+                .update("subtags", subtags.toList())
+                .addOnSuccessListener {}
```

Comment:
Question is it normal that most `addOnSuccessListener {}` are empty?

Reply:
Here you are in the function to add a category. When the operation is successful there is nothing really to return so I decided to don't have a `onSuccess` parameter for this function. So in the onSuccess there is nothing done. If you should have an `onSuccess` parameter and call it at `.addOnSuccessListener` I can change it .